src/coreclr/vm/gchelpers.cpp

```diff
@@ -183,6 +183,78 @@ inline void CheckObjectSize(size_t alloc_size)
     }
 }
 
+// TODO: what are the call sites other than the fast path helpers?
+// if there is not,
+//      isSampled =
+//          (pEEAllocContext->alloc_sampling != pAllocContext->alloc_limit);
+//
+inline Object* Alloc(ee_alloc_context* pEEAllocContext, size_t size, GC_ALLOC_FLAGS flags, bool& isSampled)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE; // returns an objref without pinning it => cooperative
+    } CONTRACTL_END;
+
+    Object* retVal = nullptr;
+    gc_alloc_context* pAllocContext = &pEEAllocContext->gc_alloc_context;
+    size_t samplingBudget = (size_t)(pEEAllocContext->alloc_sampling - pAllocContext->alloc_ptr);
+    size_t availableSpace = (size_t)(pAllocContext->alloc_limit - pAllocContext->alloc_ptr);
+    auto pCurrentThread = GetThread();
+    if (flags & GC_ALLOC_USER_OLD_HEAP)
+    {
+        // if sampling is on, decide if this object should be sampled.
+        // the provided allocation context fields are not used for LOH/POH allocations
+        // (only its alloc_bytes_uoh field will be updated),
+        // so get a random size in the distribution and if it is less than the size of the object
+        // then this object should be sampled
+        isSampled = ee_alloc_context::IsSampled(pCurrentThread->GetRandom(), size);
+    }
+    else
+    {
+        // as explained below, the GC might have returned an allocation context that is already full
+        // so, in that case, we should check if the next threshold is inside the object or not like for LOH
+        if ((pAllocContext->alloc_ptr != nullptr) && (pAllocContext->alloc_ptr == pAllocContext->alloc_limit))
```

Comment:
This code to compute isSampled looks good in the different identified scenarios (initial empty AC, no sampling AC, full AC and sampled AC within/beyond limit). 
However, for UOH, we can't rely on the value of the AC fields (i.e. Alloc() is not using nor updating them for uoh) so we need to keep it as it is (especially for POH where small objects could be allocated).

I don't understand why it is needed to deal with expectedNewXXX: why not simply always recomputing the sampling limit? In all cases, the alloc_sampling should be recomputed to be consistent so the fast path does not crash. In which scenario do you think alloc_sampling should not be recomputed after a call to Alloc()?

Reply:
There is no correctness issue if you recompute the sampling limit on every alloc. It might hurt perf but we won't know for sure until we've done some microbenchmarking.